carbonmark-api/src/routes/retirements/get.ts

```diff
@@ -0,0 +1,79 @@
+import { FastifyInstance, FastifyReply, FastifyRequest } from "fastify";
+import { GraphQLClient } from "graphql-request";
+import gql from "graphql-tag";
+
+import {
+  AllRetirementsQuery,
+  CarbonCreditFragmentFragmentDoc,
+  KlimaRetire_Filter,
+  RetireFragmentFragmentDoc,
+  Retire_Filter,
+} from "../../.generated/types/digitalCarbon.types";
+import { GRAPH_URLS } from "../../app.constants";
+import { formatRetirement } from "../../utils/helpers/retirements.utils";
+import { gqlWhereFragment } from "../../utils/helpers/utils";
+import { Querystring, schema } from "./get.schema";
+
+type FilterItem = string | number | null;
+type FilterRecords =
+  | { [k: string]: FilterItem | FilterItem[] | Filter }
+  | FilterItem
+  | FilterItem[]
+  | null;
+type Filter = FilterRecords | FilterRecords[];
+
+// Handler function for the "/retirements/klima/:account_id/:retirement_index" route
+const handler = () =>
+  async function (
+    request: FastifyRequest<{
+      Querystring: Querystring;
+    }>,
+    reply: FastifyReply
+  ) {
+    // Get graphql client
+    const network = request.query.network || "polygon";
+    const graph_urls = GRAPH_URLS[network];
+    const client = new GraphQLClient(graph_urls.digitalCarbon);
+
+    // Build filter
+    const retireFilter: Partial<
+      Omit<Retire_Filter, "klimaRetire_"> & {
+        klimaRetire_: Partial<KlimaRetire_Filter>;
+      }
+    > = {};
+    const klimaRetireFilter: Partial<KlimaRetire_Filter> | undefined = request
+      .query.retirementIndex
+      ? {
+          index: String(request.query.retirementIndex),
+        }
+      : undefined;
+    retireFilter.klimaRetire_ = klimaRetireFilter;
+    retireFilter.beneficiaryAddress = request.query.beneficiaryAddress;
+
+    // Build GQL query
+    const SearchKlimaRetirementDocument = gql`
+    query searchRetirements {
+      retires(where: ${gqlWhereFragment(retireFilter)}) {
+        ...RetireFragment
+        credit {
+          ...CarbonCreditFragment
+        }
+      }
+    }
```

Comment:
Any reason we can't add this to the `/gql` folder and interact via the sdk?

Reply:
@0xemc Because the where clause is dynamic. . I searched for ways to do it via the SDK but could not find any.
If you look at the `/projects` endpoint we have the same issue but we solve it in a convoluted way (by passing all possible values to the query arguments when we do not want to filter). There it is not possible because there is an infinite number of beneficiaryAddresses and retirement indexes.

Reply:
Hmm ok this is something that is going to come up again. We should be able to simply go:

```gql
query searchRetirements($where: Retire_filter) {
```

However I can see that the generated SDK is still just returning all results..

I'll create a ticket to address this.